File: clients/go/internal/tracejournal/dag.go
package tracejournal

import (
	"strings"

	"daisugi-verify/internal/pmodel"
	"daisugi-verify/internal/pyjson"
	"daisugi-verify/internal/pystr"
)

// Steps is a plan's steps (model_dump() objects).
func Steps(plan *pyjson.Object) []*pyjson.Object {
	xs := plan.Value("steps").([]any)
	out := make([]*pyjson.Object, len(xs))
	for i, x := range xs {
		out[i] = x.(*pyjson.Object)
	}
	return out
}

// TopoOrder is dag.topological_order: networkx's topological_sort of the
// plan's graph (nodes in step order, then each unknown dependency as its
// first edge adds it; each node's successors in the order their edges were
// added), taken generation by generation. ok is false where Python raises:
// a cycle (NetworkXUnfeasible) or a node that is not a step (KeyError).
func TopoOrder(plan *pyjson.Object) (order []*pyjson.Object, ok bool) {
	order, err := TopoOrderErr(plan)
	return order, err == nil
}

// PyError is an exception the oracle raises: its str().
type PyError struct{ Type, Msg string }

func (e *PyError) Error() string { return e.Msg }

// TopoOrderErr is TopoOrder with the exception topological_order raises:
// ValueError for a cycle, KeyError for a dependency that is not a step.
func TopoOrderErr(plan *pyjson.Object) ([]*pyjson.Object, error) {
	var order []*pyjson.Object
	steps := Steps(plan)
	byID := map[string]*pyjson.Object{}
	var nodes []string
	index := map[string]int{}
	addNode := func(n string) {
		if _, seen := index[n]; !seen {
			index[n] = len(nodes)
			nodes = append(nodes, n)
		}
	}
	for _, s := range steps {
		id := s.Value("id").(string)
		byID[id] = s
		addNode(id)
	}
	succ := map[string][]string{}
	hasEdge := map[[2]string]bool{}
	indeg := map[string]int{}
	for _, s := range steps {
		to := s.Value("id").(string)
		for _, d := range s.Value("depends_on").([]any) {
			from := d.(string)
			addNode(from)
			if hasEdge[[2]string{from, to}] {
				continue
			}
			hasEdge[[2]string{from, to}] = true
			succ[from] = append(succ[from], to)
			indeg[to]++
		}
	}
	var zero []string
	for _, n := range nodes {
		if indeg[n] == 0 {
			zero = append(zero, n)
		}
	}
	remaining := map[string]int{}
	for n, d := range indeg {
		if d > 0 {
			remaining[n] = d
		}
	}
	var ids []string
	for len(zero) > 0 {
		gen := zero
		zero = nil
		for _, n := range gen {
			for _, c := range succ[n] {
				remaining[c]--
				if remaining[c] == 0 {
					zero = append(zero, c)
					delete(remaining, c)
				}
			}
		}
		ids = append(ids, gen...)
	}
	if len(remaining) > 0 {
		return nil, &PyError{"ValueError", "Plan has a cycle; run verify(plan, envelope) before supervising"}
	}
	for _, id := range ids {
		s, isStep := byID[id]
		if !isStep {
			return nil, &PyError{"KeyError", pystr.Repr(id)}
		}
		order = append(order, s)
	}
	return order, nil
}

// StructureSignature is distiller.plan_structure_signature: the step
// types in topological order joined by an arrow. ok is false where Python
// raises (and its callers record None).
func StructureSignature(plan *pyjson.Object) (string, bool) {
	order, ok := TopoOrder(plan)
	if !ok {
		return "", false
	}
	types := make([]string, len(order))
	for i, s := range order {
		types[i] = s.Value("type").(string)
	}
	return strings.Join(types, "→"), true
}

// RefinementRecord is refinement.RefinementRecord, as far as this binary
// reads it: a step given as a dict of a registered type.
var RefinementRecord = &pmodel.Model{Name: "RefinementRecord", Fields: []pmodel.Field{
	{Name: "step", Schema: pmodel.StepField{}, Required: true},
	{Name: "violations", Schema: pmodel.List{Elem: pmodel.Violation}, Required: true},
	{Name: "z3_counterexample", Schema: pmodel.Nullable{Inner: pmodel.Dict{Val: pmodel.Any{}}}, Required: true},
	{Name: "envelope_id", Schema: pmodel.Str{}, Required: true},
	{Name: "fallback_action", Schema: pmodel.Literal{Choices: []string{"halted", "recomputed"}}, Required: true},
	{Name: "recomputed_step", Schema: pmodel.StepField{Nullable: true}, Default: func() any { return nil }},
	{Name: "recomputed_verification", Schema: pmodel.Nullable{Inner: pmodel.VerificationResult}, Default: func() any { return nil }},
	{Name: "timestamp", Schema: pmodel.Float{}, Required: true},
	{Name: "cache_key", Schema: pmodel.Nullable{Inner: pmodel.Str{}}, Default: func() any { return nil }},
}}

File: clients/rust/src/tracejournal/dag.rs
//! `dag.topological_order` and `distiller.plan_structure_signature` over
//! a plan's `model_dump()`.

use std::collections::HashMap;

use crate::gate::py::text::repr;
use crate::gate::pyjson::{Object, Value};

/// An exception the oracle raises: its type and str().
#[derive(Debug, Clone)]
pub struct PyError {
    pub typ: &'static str,
    pub msg: String,
}

/// A plan's steps.
pub fn steps(plan: &Object) -> Vec<&Object> {
    match plan.value("steps") {
        Value::List(l) => l.iter().filter_map(|s| s.as_obj()).collect(),
        _ => vec![],
    }
}

fn deps(s: &Object) -> Vec<String> {
    match s.value("depends_on") {
        Value::List(l) => l.iter().filter_map(|d| d.as_str().map(|x| x.to_string())).collect(),
        _ => vec![],
    }
}

/// `dag.topological_order`: networkx's topological_sort of the plan's
/// graph (nodes in step order, then each unknown dependency as its first
/// edge adds it; each node's successors in the order their edges were
/// added), taken generation by generation. The error is the one Python
/// raises: ValueError for a cycle, KeyError for a dependency that is not
/// a step.
pub fn topo_order(plan: &Object) -> Result<Vec<&Object>, PyError> {
    let ss = steps(plan);
    let mut by_id: HashMap<String, &Object> = HashMap::new();
    let mut nodes: Vec<String> = vec![];
    let mut index: HashMap<String, usize> = HashMap::new();
    let add = |n: &str, nodes: &mut Vec<String>, index: &mut HashMap<String, usize>| {
        if !index.contains_key(n) {
            index.insert(n.to_string(), nodes.len());
            nodes.push(n.to_string());
        }
    };
    for s in &ss {
        let id = s.value("id").as_str().unwrap_or("").to_string();
        by_id.insert(id.clone(), s);
        add(&id, &mut nodes, &mut index);
    }
    let mut succ: HashMap<String, Vec<String>> = HashMap::new();
    let mut edges: std::collections::HashSet<(String, String)> = Default::default();
    let mut indeg: HashMap<String, usize> = HashMap::new();
    for s in &ss {
        let to = s.value("id").as_str().unwrap_or("").to_string();
        for from in deps(s) {
            add(&from, &mut nodes, &mut index);
            if !edges.insert((from.clone(), to.clone())) {
                continue;
            }
            succ.entry(from).or_default().push(to.clone());
            *indeg.entry(to.clone()).or_default() += 1;
        }
    }
    let mut zero: Vec<String> = nodes.iter().filter(|n| indeg.get(*n).copied().unwrap_or(0) == 0).cloned().collect();
    let mut remaining: HashMap<String, usize> = indeg.into_iter().filter(|(_, d)| *d > 0).collect();
    let mut ids: Vec<String> = vec![];
    while !zero.is_empty() {
        let gen = std::mem::take(&mut zero);
        for n in &gen {
            if let Some(cs) = succ.get(n) {
                for c in cs {
                    if let Some(d) = remaining.get_mut(c) {
                        *d -= 1;
                        if *d == 0 {
                            remaining.remove(c);
                            zero.push(c.clone());
                        }
                    }
                }
            }
        }
        ids.extend(gen);
    }
    if !remaining.is_empty() {
        return Err(PyError { typ: "ValueError", msg: "Plan has a cycle; run verify(plan, envelope) before supervising".into() });
    }
    let mut order = vec![];
    for id in ids {
        match by_id.get(&id) {
            Some(s) => order.push(*s),
            None => return Err(PyError { typ: "KeyError", msg: repr(&id) }),
        }
    }
    Ok(order)
}

/// `distiller.plan_structure_signature`: the step types in topological
/// order joined by an arrow; None where Python raises.
pub fn structure_signature(plan: &Object) -> Option<String> {
    let order = topo_order(plan).ok()?;
    Some(order.iter().map(|s| s.value("type").as_str().unwrap_or("")).collect::<Vec<_>>().join("\u{2192}"))
}
